packages/dev/core/src/Collisions/gpuPicker.ts

```diff
@@ -0,0 +1,224 @@
+            Constants.TEXTURE_NEAREST_NEAREST
+        );
+    }
+
+    private _createColorMaterial(scene: Scene) {
+        if (this._renderMaterial) {
+            this._renderMaterial.dispose();
+        }
+
+        const defines: string[] = [];
+        const options = {
+            attributes: [VertexBuffer.PositionKind],
+            uniforms: ["world", "viewProjection", "color"],
+            needAlphaBlending: false,
+            defines: defines,
+            useClipPlane: null,
+        };
+
+        this._renderMaterial = new ShaderMaterial("colorShader", scene, "color", options, false);
+
+        const callback = (mesh: AbstractMesh | undefined) => {
+            if (!mesh) {
+                return;
+            }
+
+            const effect = this._renderMaterial!.getEffect();
+
+            effect.setColor4("color", this._idColors[mesh.uniqueId], 1);
+
+            this._meshRenderingCount++;
+        };
+
+        this._renderMaterial.customBindingObservable.add(callback);
+    }
+
+    /**
+     * Execute a picking operation
+     * @param x defines the X coordinates where to run the pick
+     * @param y defines the Y coordinates where to run the pick
+     * @param scene defines the scene to pick from
+     * @param disposeWhenDone defines a boolean indicating we do not want to keep resources alive
+     * @returns A promise with the picking results
+     */
+    public pickAsync(x: number, y: number, scene: Scene, disposeWhenDone = true): Promise<Nullable<Mesh>> {
+        const engine = scene.getEngine();
+        const rttSizeW = engine.getRenderWidth();
+        const rttSizeH = engine.getRenderHeight();
+
+        if (!this._readbuffer) {
+            this._readbuffer = new Uint8Array(engine.isWebGPU ? 256 : 4); // Because of block alignment in WebGPU
+        }
+
+        this._meshRenderingCount = 0;
+        // Ensure ints
+        x = x >> 0;
+        y = y >> 0;
+
+        if (x < 0 || y < 0 || x >= rttSizeW || y >= rttSizeH) {
+            return Promise.resolve(null);
+        }
+
+        // Invert Y
+        y = rttSizeH - y;
+
+        if (!this._pickingTexure) {
+            this._createRenderTarget(scene, rttSizeW, rttSizeH);
+        } else {
+            const size = this._pickingTexure.getSize();
+
+            if (size.width !== rttSizeW || size.height !== rttSizeH || this._cachedScene !== scene) {
+                this._pickingTexure.dispose();
+                this._createRenderTarget(scene, rttSizeW, rttSizeH);
+            }
+        }
+
+        if (!this._cachedScene || this._cachedScene !== scene) {
+            this._createColorMaterial(scene);
+        }
+
+        this._cachedScene = scene;
+        scene.customRenderTargets.push(this._pickingTexure!);
+        this._pickingTexure!.renderList = [];
+        this._pickingTexure!.clearColor = new Color4(0, 0, 0, 0);
+
+        // We need to give every mesh an unique color
+        this._pickingTexure!.onBeforeRender = () => {

```

Comment:
just a nit but would be great to move the functions out to prevent recreating them on every pick ?

Reply:
it is a bit of a mess then to store all the resolve, reject, x, y etc... :(